### PR TITLE
HPCC4J-577 Added Read Retry to HPCCRemoteFileReader

### PR DESCRIPTION
**Pull request by jpmcmu**
- Added Read Retry to HPCCRemoteFileReader
- Minor improvements to FileUtility to improve testing

Signed-off-by: James McMullan James.McMullan@lexisnexis.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [X] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [X] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [ ] This change does not cause any existing JUnits to fail.
  - [ ] I have include JUnit coverage to test this change
  - [ ] I have performed system test and covered possible regressions and side effects.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] This change fixes the problem, not just the symptom

### CODE REVIEW

**Comment by jpmcmu** [comment]
@drealeed Thanks for the code review! Implemented code review changes. @rpastrana please review

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java [hunk 1]
```diff
@@ -425,6 +426,7 @@ private static Options getReadTestOptions()
         options.addOption("user", true, "Specifies the username used to connect. Defaults to null.");
         options.addOption("pass", true, "Specifies the password used to connect. Defaults to null.");
         options.addOption("num_threads", true, "Specifies the number of parallel to use to perform operations.");
+        options.addOption("expiry_seconds", true, "Access token expiration seconds.");
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java:429]: access_expiry_secs ?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/HpccRemoteFileReader.java [hunk 2]
```diff
@@ -189,18 +195,23 @@ public HpccRemoteFileReader(DataPartition dp, FieldDef originalRD, IRecordBuilde
...
         {
             connectTimeout = RowServiceInputStream.DEFAULT_CONNECT_TIMEOUT_MILIS;
         }
+        this.connectTimeout = connectTimeout;
 
-        this.dataPartition = dp;
-        this.recordBuilder = recBuilder;
+        if (this.originalRecordDef == null)
+        {
+            throw new Exception("HpccRemoteFileReader: Original record definition is null.");
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/HpccRemoteFileReader.java:213]: is there any guidance we can provide to the user in the err message?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/HpccRemoteFileReader.java [hunk 3]
```diff
@@ -246,6 +257,51 @@ public HpccRemoteFileReader(DataPartition dp, FieldDef originalRD, IRecordBuilde
         openTimeMs = System.currentTimeMillis();
     }
 
+    private boolean retryRead()
+    {
+        if (retryCount < 3)
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/HpccRemoteFileReader.java:262]: shouldn't the max retry count be configurable?

Comment (jpmcmu) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/HpccRemoteFileReader.java]: @rpastrana Implemented code review changes, will squash if it looks good

**Comment by jpmcmu** [comment]
@rpastrana squashed

**Pull request merged**